nssdb/keys.go

```diff
@@ -0,0 +1,208 @@
+package nssdb
+
+import (
+	"context"
+	"crypto/ecdh"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"encoding/asn1"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"golang.org/x/crypto/cryptobyte"
+)
+
+// The only supported params for now
+var ecParams []byte
+
+func init() {
+	ecParams, _ = hex.DecodeString("06082a8648ce3d030107")
+}
+
+func (obj Object) ToECPublicKey() (*ecdsa.PublicKey, error) {
+	if err := obj.ValidateULong("CKA_CLASS", CKO_PUBLIC_KEY); err != nil {
+		return nil, err
+	}
+	if err := obj.ValidateULong("CKA_KEY_TYPE", CKK_EC); err != nil {
+		return nil, err
+	}
+	// TODO(areed) actually parse the params
+	if err := obj.Validate("CKA_EC_PARAMS", ecParams); err != nil {
+		return nil, err
+	}
+
+	ecPointASN1, ok := obj.Attributes["CKA_EC_POINT"]
+	if !ok {
+		return nil, errors.New("object is missing attribute CKA_EC_POINT")
+	}
+	var ecPoint []byte
+	rest, err := asn1.Unmarshal(ecPointASN1, &ecPoint)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode CKA_EC_POINT: %w", err)
+	} else if len(rest) > 0 {
+		return nil, errors.New("failed to decode CKA_EC_POINT")
+	}
+
+	// TODO(areed) parse curve from CKA_EC_PARAMS
+	pk, err := ecdh.P256().NewPublicKey(ecPoint)
```

Comment:
Ok, this will probably fail if the `ecPoint` is from a P-384 or P-512 curve. We could validate the size before.

Reply:
We validate on line 27 that it is P-256 according to CKA_EC_PARAMS